provider_gsuite.go

```diff
@@ -0,0 +1,217 @@
+			"G Suite user with permission to access the Admin APIs")
+	}
+	if config.GroupsRecurseMaxDepth < 0 {
+		return errors.New("'gsuite_recurse_max_depth' must be a positive integer")
+	}
+
+	// Create the google JWT config from the service account key file
+	if g.jwtConfig, err = google.JWTConfigFromJSON(config.serviceAccountKeyJSON,
+		admin.AdminDirectoryGroupReadonlyScope, admin.AdminDirectoryUserReadonlyScope); err != nil {
+		return err
+	}
+
+	// Set the subject to impersonate and config
+	g.jwtConfig.Subject = config.AdminImpersonateEmail
+	g.config = config
+	return nil
+}
+
+// SensitiveKeys returns keys that should be redacted when reading the config of this provider
+func (g *GSuiteProvider) SensitiveKeys() []string {
+	return []string{}
+}
+
+// FetchGroups fetches and returns groups from G Suite.
+func (g *GSuiteProvider) FetchGroups(b *jwtAuthBackend, allClaims map[string]interface{}, role *jwtRole) (interface{}, error) {
+	if !g.config.FetchGroups {
+		return nil, nil
+	}
+
+	userName, err := g.getUserClaim(b, allClaims, role)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set context and create a new admin service for requests to Google admin APIs
+	g.ctx = b.providerCtx
+	g.adminSvc, err = admin.NewService(g.ctx, option.WithHTTPClient(g.jwtConfig.Client(g.ctx)))
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the G Suite groups
+	userGroupsMap := make(map[string]bool)
+	if err := g.search(userGroupsMap, userName, g.config.GroupsRecurseMaxDepth); err != nil {
+		return nil, err
+	}
+
+	// Convert set of groups to list
+	var userGroups = make([]interface{}, 0, len(userGroupsMap))
+	for email := range userGroupsMap {
+		userGroups = append(userGroups, email)
+	}
+
+	b.Logger().Debug("fetched G Suite groups", "groups", userGroups)
+	return userGroups, nil
+}
+
+// search recursively searches for G Suite groups based on a configured depth for this provider.
+func (g *GSuiteProvider) search(visited map[string]bool, userName string, depth int) error {
+	var newGroups []string
+	call := g.adminSvc.Groups.List().UserKey(userName).Fields("nextPageToken", "groups(email)")
+	if err := call.Pages(g.ctx, func(groups *admin.Groups) error {
+		for _, group := range groups.Groups {
+			if _, ok := visited[group.Email]; ok {
+				continue
+			}
+			visited[group.Email] = true
+			newGroups = append(newGroups, group.Email)
+		}
+		// Only recursively search for new groups that haven't been seen
+		if depth > 0 {
+			for _, email := range newGroups {
+				if err := g.search(visited, email, depth-1); err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// FetchUserInfo fetches additional user information from G Suite using custom schemas.
+func (g *GSuiteProvider) FetchUserInfo(b *jwtAuthBackend, allClaims map[string]interface{}, role *jwtRole) error {
+	if !g.config.FetchUserInfo || g.config.UserCustomSchemas == "" {
```

Comment:
Maybe we could log a warning if UserCustomSchemas != "" and FetchUserInfo is left to false (the default value)?

Reply:
+1 to adding a warning

Reply:
Done in c1c84bd05e70cec387bee277af31f07ed47ef35f 👍